File: AlgorithmsPackCPP/343. Integer Break.cpp
const int MAX = 58+1;
int memory[MAX];

class Solution {
public:
    int intBreak(int n){
        if(n==1)
            return 1;
        
        auto&ret = memory[n];
        if(ret!=-1)
            return ret;
        
        ret = n;
        for(int i = 1; i < n; i++){
            ret = max(ret,i*intBreak(n-i));
        }
        return ret;
    }
    
    int integerBreak(int n) {
        if(n==2||n==3)
            return n-1;
        
        memset(memory,-1,sizeof(memory));
        return intBreak(n);
    }
};

File: AlgorithmsPackPythonSolutions/343. Integer Break.py
class Solution:
    def integerBreak(self, n: int) -> int:
        MAX = 58 + 1
        memory = [-1] * MAX

        def intBreak(num):
            if num == 1:
                return 1

            if memory[num] != -1:
                return memory[num]

            result = num
            for i in range(1, num):
                result = max(result, i * intBreak(num - i))

            memory[num] = result
            return result

        if n == 2 or n == 3:
            return n - 1

        return intBreak(n)
